### PR TITLE
Added support of default implementation in trait definition

### PR DESCRIPTION
**Pull request by xgreenx**
Updated contract macros.
It allows defining default implementation in external trait + allows define not external messages(without `#[ink(message]` attribute).
Now it copies implementation from trait definition and pastes implementation in impl section directly.
But it means that we copy the implementation of the child contract directly into the current contract.
It requires us to make all parts of implementation public and import them like `contract::traits::*;`. 
But, it simplifies the code usage.
It also adds a feature to call(paste) the implementation from the trait(some kind of super call).

### CODE REVIEW

**Review by VargSupercolony — APPROVED** [review]
a lot of changes: added macros, reworked code according to trait logic change, updated READMEs. please add comments to macros as soon as possible :)

File: contracts/brush/proc_macros/contract.rs
Comment (VargSupercolony) [inline contracts/brush/proc_macros/contract.rs:80]: the logic is identical for storage_trait and trait_definition?

File: contracts/token/psp721/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp721/traits.rs:151]: maybe create a `pub(crate) const ZERO_ADDRESS: [u8; 32] = [0u8; 32]` and reuse?

File: examples/ownable/README.md
Comment (VargSupercolony) [inline examples/ownable/README.md:74]: why use burn example on L73? we're implementing mint

File: examples/ownable/lib.rs
Comment (VargSupercolony) [inline examples/ownable/lib.rs:32]: change // #[super]self.burn(...) to #[super]self.mint(...)

**Review by VargSupercolony — CHANGES_REQUESTED** [review]
Great work, comments are helpful for understanding macros. Also now the library is much more usable with default implementations inline. Again, great!

File: contracts/brush/proc_macros/contract.rs
Comment (xgreenx) [reply contracts/brush/proc_macros/contract.rs:80]: Yes and no, in both cases we need to extract attribute, but in one case - trait_definition, in another - storage_trait

Comment (VargSupercolony) [reply contracts/brush/proc_macros/contract.rs:80]: ok I see, no need to overenginner with extracting 🙃 it's all good

File: contracts/token/psp721/traits.rs
Comment (xgreenx) [reply contracts/token/psp721/traits.rs:151]: No, we can't do that, because it must be public then(After this change whole implementation of some contracts must be public because we copy implementation from one crate to another). 
We can try to use `ZERO_ADDRESS`, but I don't know how it will resolve conflicts with `ZERO_ADDRESS` in another crates=)

Comment (VargSupercolony) [reply contracts/token/psp721/traits.rs:151]: Will using this const as `crate::ZERO_ADDRESS` work fine?  I guess then won't conflict with other `ZERO_ADDRESS`es

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:43]: Do we need this line in doc? I don't think it's necessary for public usage. I guess that mentioning that  `#[brush::trait_definition]` should be defined in `#[brush::contract]` should be enough

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:160]: (e.g. given `_method_name_mut` function, field `method_name` will be generated)

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:161]: minor: missing space  ...getter(It m...

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:173]: The same as above. I think it's more implementation details, than usage. Let's just say that this macro should be inside bruch::contract

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:44]: minor: . in the end

File: contracts/brush/proc_macros/lib.rs
Comment (sventime) [inline contracts/brush/proc_macros/lib.rs:174]: minor: . in the end

File: contracts/token/psp1155/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp1155/traits.rs:16]:
```suggestion
// We don't need to expose it, because ink! will define AccountId, Balance and StaticEnv itself.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:11]:
```suggestion
// We don't need to expose it, because ink! will define AccountId, Balance and StaticEnv itself.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:97]:
```suggestion
    /// Panics with `InsufficientBalance` error if there are not enough tokens on
    /// the caller's account Balance.
    ///
    /// Panics with `ZeroSenderAddress` error if sender's address is zero.
    ///
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:129]:
```suggestion
    /// Panics with `InsufficientAllowance` error if there are not enough tokens allowed
    /// for the caller to withdraw from `from`.
    ///
    /// Panics with `InsufficientBalance` error if there are not enough tokens on
    /// the the account Balance of `from`.
    ///
    /// Panics with `ZeroSenderAddress` error if sender's address is zero.
    ///
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:150]:
```suggestion
    /// Panics with `ZeroSenderAddress` error if sender's address is zero.
    ///
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:165]:
```suggestion
    /// Panics with `ZeroSenderAddress` error if sender's address is zero.
    ///
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:183]:
```suggestion
    /// Panics with `InsufficientAllowance` error if there are not enough tokens allowed
    /// by owner for `spender`.
    ///
    /// Panics with `ZeroSenderAddress` error if sender's address is zero.
    ///
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:198]:
```suggestion
    // Helper functions

    /// Emit transfer event. It must be implemented in inheriting struct
    fn _emit_transfer_event(&self, _from: Option<AccountId>, _to: Option<AccountId>, _amount: Balance) {}

    /// Emit approval event. It must be implemented in inheriting struct
```

Comment (VargSupercolony) [reply contracts/token/psp20/traits.rs:198]: the trait does not inherit the struct, a struct inherits from a trait 🙃

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:231]:
```suggestion
    /// Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

File: contracts/token/psp20/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp20/traits.rs:251]:
```suggestion
    /// Panics with `ZeroSenderAddress` error if recipient's address is zero.
    ///
    /// Panics with `InsufficientBalance` error if there are not enough tokens on
    /// the account balance of `account`.
```

File: contracts/token/psp721/traits.rs
Comment (VargSupercolony) [inline contracts/token/psp721/traits.rs:13]:
```suggestion
// We don't need to expose it, because ink! will define AccountId and StaticEnv itself.
```

File: examples/ownable/README.md
Comment (VargSupercolony) [inline examples/ownable/README.md:4]:
```suggestion
[psp1155](contracts/token/psp1155) together to provide `only owner` rights to mint and burn tokens.
```

**Review by VargSupercolony — APPROVED** [review]
great, approve from me after merging spelling error suggestions

**Review by sventime — APPROVED** [review]

**Pull request merged**
